### PR TITLE
consolidate json settings

### PR DESCRIPTION
**Pull request by tomerd**
motivation: consolidated logic

changes:
* new factory functions for JSONEncoder and JSONDecoder with healthy defaults
* refactor code to use new factories
* update CMake config where required to bring in Basics

### CODE REVIEW

**Review by yim-lee — APPROVED** [review]
Looks great 👍

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
infra failure. @swift-ci please smoke test linux

**Review by neonichu — APPROVED** [review]

**Review by abertelrud — APPROVED** [review]
Looks good to me.  I'm not over-the-moon about the "makeWith..." names, but don't have any better suggestions.  Ideally it would be some kind of convenience initializer, but that's not really possible without having parameters, I think.

**Comment by tomerd** [comment]
> Looks good to me. I'm not over-the-moon about the "makeWith..." names, but don't have any better suggestions. Ideally it would be some kind of convenience initializer, but that's not really possible without having parameters, I think.

same, happy to rename if someone has a suggestion

**Comment by abertelrud** [comment]
> > Looks good to me. I'm not over-the-moon about the "makeWith..." names, but don't have any better suggestions. Ideally it would be some kind of convenience initializer, but that's not really possible without having parameters, I think.

> same, happy to rename if someone has a suggestion

Here's a thought:  the reason we can't have a convenience initializer is that we don't have a parameter, but it occurs to me that we actually have one: the date formatting strategy.  The other formatting options in the encoder are just eye candy, and don't change the validity of the document.  But AFAIK the date formatting strategy does affect the semantics of the document (which is also, I suspect, why it needs to be set in the decoder).

Maybe it's a bit of a cop-out, but could we have a parameter that says what date encoding strategy to use?  Something like `JSONEncoder(withDateEncodingStrategy:)` and `JSONDecoder(withDateDecodingStrategy:)`?  Maybe a bit of a cop-out but it would highlight the use of ISO 8601 at the call site, which seems semantically meaningful, and provide a way of spelling this is a convenience initializer.

**Comment by tomerd** [comment]
> Maybe it's a bit of a cop-out, but could we have a parameter that says what date encoding strategy to use? Something like JSONEncoder(withDateEncodingStrategy:) and JSONDecoder(withDateDecodingStrategy:)? Maybe a bit of a cop-out but it would highlight the use of ISO 8601 at the call site, which seems semantically meaningful, and provide a way of spelling this is a convenience initializer. There would need to be some special case about .customISO8601 when .iso8601 isn't available, but that seems like implementation details.

I dont mind making this change, but I will say that ISO8601 is pretty much the default for dates in JSON, so likely to always be needed.

**Comment by abertelrud** [comment]
> I dont mind making this change, but I will say that ISO8601 is pretty much the default for dates in JSON, so likely to always be needed.

I see — I didn't realize that.  In that case perhaps it should be just left as is until someone comes up with a better idea.

**Comment by tomerd** [comment]
> I see — I didn't realize that. In that case perhaps it should be just left as is until someone comes up with a better idea.

made one change to give the call site more control over the behavior

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Review by abertelrud — APPROVED** [review]

**Review by abertelrud — APPROVED** [review]

File: Sources/Basics/JSON+Extensions.swift [hunk 1]
```diff
@@ -39,3 +39,64 @@ extension JSONDecoder.DateDecodingStrategy {
...
+            #if os(macOS) || os(iOS) || os(tvOS) || os(watchOS)
+            if #available(macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0, *) {
+                encoder.outputFormatting = [.sortedKeys, .prettyPrinted, .withoutEscapingSlashes]
+            } else if #available(macOS 10.13, iOS 11.0, watchOS 4.0, tvOS 11.0, *) {
+                encoder.outputFormatting = [.sortedKeys, .prettyPrinted]
+            } else {
+                encoder.outputFormatting = [.prettyPrinted]
+            }
+            #elseif compiler(>=5.3)
+            encoder.outputFormatting = [.sortedKeys, .prettyPrinted, .withoutEscapingSlashes]
```
Comment (liuliu) [inline Sources/Basics/JSON+Extensions.swift:93]: `.withoutEscapingSlashes` is not in 5.3 on non-Darwin platforms.

Comment (yim-lee) [reply Sources/Basics/JSON+Extensions.swift:93]: It's in the latest 5.3.2 I believe.

Comment (liuliu) [reply Sources/Basics/JSON+Extensions.swift:93]: Thanks! Yep, verified.

**Pull request merged**